<a href="https://colab.research.google.com/github/Kanka-goswami/IISc_CDS_2302064/blob/main/27Jan'24_M1_NB_MiniProject_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Advanced Certification Program in Computational Data Science
## A program by IISc and TalentSprint
### Mini Project Notebook: Linear Algebra and Calculus

## Problem Statement

 The task is to advise a petroleum company on how to meet the demands of their customers for motor oil, diesel oil and gasoline.

## Learning Objectives

At the end of the experiment, you will be able to

* create arrays and matrices in python
* understand the concepts of linear equations
* solve the system of linear equations

### Data

From a barrel of crude oil, in one day, factory $A$ can produce
* 20 gallons of motor oil,
* 10 gallons of diesel oil, and
* 5 gallons of gasoline

Similarly, factory $B$ can produce
* 4 gallons of motor oil,
* 14 gallons of diesel oil, and
* 5 gallons of gasoline

while factory $C$ can produce
* 4 gallons of motor oil,
* 5 gallons of diesel oil, and
* 12 gallons of gasoline

There is also waste in the form of paraffin, among other things. Factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude, factory $B$ 5 gallons, and factory $C$ 2 gallons.

**Note:** Your conclusion should include a discussion of the nature of the terms *unique*, *no solution*, *overdetermined* and *underdetermined* as they apply in the context of the oil plants.

## Grading = 10 Points

### Create an array

Create an array of size 2x3 with arbitrary values.

In [64]:
# YOUR CODE HERE
import numpy as np
a = np.array([[1, 2, 3],
              [4, 5, 6]])
a

array([[1, 2, 3],
       [4, 5, 6]])

### Create the system of Linear Equations

Suppose the current daily demand from distributors is 6600 gallons of motor oil, 5100 gallons of diesel oil and 3100 of gasoline.

Set up the system of equations which describes the above situation. Please include the units as well.

Let the number of barrels used by factory $A$, $B$ and $C$ are $x$, $y$ and $z$ respectively.

Then the system of linear equations will be

$$Motor\ oil:\ \ \ 20x + 4y + 4z = 6600$$

$$Diesel\ oil:\ \ \ 10x + 14y + 5z = 5100$$

$$Gasoline:\ \ \ 5x + 5y + 12z = 3100$$

### Solve the system of Linear Equation (2 points)

How many barrels of crude oil each plant should get in order to meet the demand as a group. Remember that we can only provide each plant with an integral number of barrels.

In [65]:
# YOUR CODE HERE
from scipy.linalg import solve
A = np.array([[20, 4, 4],
              [10, 14, 5],
              [5, 5, 12]])
b = np.array([6600, 5100, 3100])
X1 = solve(A,b)
print(X1)
print(np.ceil(X1))

[287.25 128.75  85.  ]
[288. 129.  85.]


**INFERENCE:** *Therefore, the number of barrels required by factory A,B and C are 288, 129 and 85 respectively*

Suppose the total demand for all products **doubled**. What would the solution now be? How does it compare to the original solution? Why, mathematically, should this have been expected?

In [66]:
# YOUR CODE HERE
A = np.array([[20, 4, 4],
              [10, 14, 5],
              [5, 5, 12]])
b = np.array([6600 * 2, 5100 * 2, 3100 * 2])
X2 = solve(A,b)
X2

array([574.5, 257.5, 170. ])

**INFERENCE:** *From above it can be seen that as the demand doubled the number of barrels required also gets doubled.*



- Demand column $b$  if **doubled**  
$$
b' = 2*b \\
\Rightarrow 2*\textbf{A}x = b' \\
⇒ \textbf{A}(2*x) = b' \\
⇒ x' = 2*x;  \text{where $x'$ is the new solution}
$$

Suppose that the company acquires another group of distributors and that the daily demand of this group is 2000 gallons of motor oil, 4000 gallons of gasoline, and 4000 gallons of diesel oil. How would you set up production of just this supply? Are there any options (more than one way)?

In [67]:
# YOUR CODE HERE
A = np.array([[20, 4, 4],
              [10, 14, 5],
              [5, 5, 12]])
b = np.array([2000, 4000, 4000])
X3 = solve(A,b)
X3

array([ 12.5, 187.5, 250. ])

**INFERENCE:** *From above it can be seen that for the demand (2000, 4000, 4000), the barrel requirement for factories A, B & C are 13, 188 and 250 respectively. As it is a unique solution, this is the only way to fulfill the demand*

Next, calculate the needs of each factory (in barrels of crude, as usual) to meet the total demand of both groups of distributors. When you have done this, compare your answer to results already obtained. What mathematical conclusion can you draw?

In [68]:
# YOUR CODE HERE
A = np.array([[20, 4, 4],
              [10, 14, 5],
              [5, 5, 12]])
b = np.array([6600 + 2000, 5100 + 4000, 3100 + 4000])
X4 = solve(A,b)
X4

array([299.75, 316.25, 335.  ])

In [69]:
X1 + X3 == X4

array([ True,  True,  True])

**INFERENCE:** *From above we can conclude that the number of barrels required by the plants to meet the total demand (6600 + 2000, 5100 + 4000, 3100 + 4000) is equal to the sum of number of barrels required by the plants to meet the demands individually*

### Sensitivity and Robustness (1 point)

In real life applications, constants are rarely ever exactly equal to their stated value; certain amounts of uncertainty are always present. This is part of the reason for the science of statistics. In the above model, the daily productions for the plants would be averages over a period of time. Explore what effect small changes in the parameters have on the output.

To do this, pick any 3 coefficients, one at a time, and increase or decrease them by 3%. For each case , note what effect this has on the solution, as a percentage change. Can you draw any overall conclusion?

###### INCREASE BY 3% (FACTORY A)

In [70]:
# TO Review
# Increasing the coefficient value for factory A by 3%
A = np.array([[20* 1.03, 4, 4],
              [10 * 1.03, 14, 5],
              [5 *1.03, 5, 12]])

b = np.array([6600, 5100, 3100])
X5 = solve(A,b)
X5


array([278.88349515, 128.75      ,  85.        ])

In [71]:
# Percentage change in solution
(X5 - X1) / X1

array([-0.02912621,  0.        ,  0.        ])

Change in solution will be -Δp/(1 + Δp), where Δp is fraction change in coefficients of factory; assuming all are changed by same fraction

**INFERENCE:** *From above it can be seen that after increasing the coefficients of factory A  by 3%, barrel requirement for factory A has decreased by around 3% (-0.029) while for factory B & C it is constant (0).*

###### DECREASE BY 3% FACTORY A

In [72]:
# Decreasing the coefficient value for factory A by 3%
A = np.array([[20 * 0.97, 4, 4],
              [10 * 0.97, 14, 5],
              [5 * 0.97, 5, 12]])

b = np.array([6600, 5100, 3100])
X6 = solve(A,b)
X6

array([296.13402062, 128.75      ,  85.        ])

In [73]:
# Percentage change in solution
(X6 - X1) / X1

array([0.03092784, 0.        , 0.        ])

**INFERENCE**: *From above it can be seen that after DECREASING the coefficients of factory A by 3%, barrel requirement for factory A has INCREASED by around 3% (0.03) while for factory B & C it is constant (0).*

###### SIMILARY INC & DEC BY 3% (FACTORY B)

In [110]:
# Increasing the coefficient value for factory B by 3%
A = np.array([[20, 4 , 4],
              [10 * 1.03, 14, 5],
              [5, 5 , 12]])
condition_number8 = np.linalg.cond(A,p=1)
print(condition_number8)
b = np.array([6600, 5100, 3100])
X7 = solve(A,b)
X7

4.92393376818866


array([288.25890617, 120.82288008,  87.88258906])

In [111]:
# Percentage change in solution
(X7 - X1) / X1

array([ 0.00351229, -0.06156986,  0.03391281])

In [76]:
# Decreasing the coefficient value for factory B by 3%
A = np.array([[20, 4 * 0.97, 4], [10, 14 * 0.97, 5], [5, 5 * 0.97, 12]])
b = np.array([6600, 5100, 3100])
condition_number8 = np.linalg.cond(A,p=1)
print(condition_number8)
X8 = solve(A,b)
X8

4.927061855670104


array([287.25      , 132.73195876,  85.        ])

In [77]:
# Checking impact of change
p = 1e-8
A = np.array([[20 + p, 4, 4], [10, 14, 5-p], [5, 5, 12]])
b = np.array([6600, 5100, 3100])
condition_number9 = np.linalg.cond(A)
print(condition_number8)
X9 = solve(A,b)
X9

4.927061855670104


array([287.24999982, 128.75000019,  85.        ])

In [78]:
# Percentage change in solution
(X9 - X1) / X1

array([-6.30356213e-10,  1.48830927e-09, -5.17156409e-11])

**INFERENCE:** *The results are similar for factory B as well.*

**FINAL INFERENCE:** *So, we conclude that if we increase the value of coefficient by n% then its output is decreased by approximately n% and vice-versa. INVERSE PROPORTIONALITY. Also, the solution of the problem seem robust to noises in the coefficients, also inferred from condition number of the coefficient matrix*

### A Plant Off-Line (1 point)

Suppose factory $C$ is shut down by the EPA (Environmental Protection Agency) temporarily for excessive emissions into the atmosphere. If your demand is as it was originally (6600, 5100, 3100), what would you now say about the companies ability to meet it? What do you recommend they schedule for production now?

**DEDUCTION:** *So now, the problem statement chnages a bit.Let the number of barrels used by factory A & B are x & y respectively*
*  20x + 4y = 6600
*  10x + 14y = 5100
*  5x + 5y = 3100

**INFERENCE:** *The above system has 3 equations and 2 unknowns (x and y), which is* overdetermined *because number of linearly independent equations in the system are more than the unknowns*

In [79]:
# YOUR CODE HERE
# For the first and second equations
A = np.array([[20, 4], [10, 14]])
b = np.array([6600, 5100])
X9 = solve(A,b)
X9

array([300., 150.])

In [80]:
# For the first and third equations
A = np.array([[20, 4], [5, 5]])
b = np.array([6600, 3100])
X10 = solve(A,b)
X10

array([257.5, 362.5])

In [81]:
# For the second and third equations
A = np.array([[10, 14], [5, 5]])
b = np.array([5100, 3100])
X11 = solve(A,b)
X11

array([ 895., -275.])

**INFERENCE:** *There is one solution for each pair of linear equations:*

*   first and second equations (300, 150)
*   first and third (258, 363)
*   the second and third (895, -275)

*We can discard the third one as number of barrels can't be negative. However, there is no exact solution that satisfies all three simultaneously*

For over-determined system, closest solution can be estimated by solving normal equations; AᵀX = AᵀB

In [82]:
motor_oil_equation4 = np.array([20, 4])
diesel_oil_equation4 = np.array([10, 14])
gasoline_equation4 = np.array([5, 5])
demand_vector = np.array([6600, 5100, 3100])
coefficient_matrix_c_off = np.array([motor_oil_equation4, diesel_oil_equation4, gasoline_equation4])
newCoeff = np.dot(coefficient_matrix_c_off.T,coefficient_matrix_c_off)
newBvec = np.dot(coefficient_matrix_c_off.T,demand_vector)

In [83]:
solution = solve(newCoeff, newBvec)
print ('Crude oil required by factor A and B in order to closely match the demand', solution)
print('\nDemand that can be met from the two factories for motor oil, diesel, gasoline, respecively is:', np.dot(coefficient_matrix_c_off,np.ceil(solution)))

Crude oil required by factor A and B in order to closely match the demand [299.47204969 168.47826087]

Demand that can be met from the two factories for motor oil, diesel, gasoline, respecively is: [6676. 5366. 2345.]


$$ \text{Gasoline falling short by} = 3100-2340 = 760 \text{ gallons} \\
\Rightarrow 5*x+5*y = 760 \text{ and } \frac{x}{y} = \frac{300}{168} \\
\Rightarrow x = 97 \text{ barrels};y=55 \text{ barrels} \\
\therefore \left[ \begin{matrix} x \\ y \end{matrix} \right] = \left[ \begin{matrix} 397 \\ 223 \end{matrix} \right]\\
$$

In [84]:
print('Demand matrix= ','\n',demand_vector,'\n')
print('Checking the solution \n',np.dot(coefficient_matrix_c_off,[397,223]))

Demand matrix=  
 [6600 5100 3100] 

Checking the solution 
 [8832 7092 3100]


### Buying another plant

####(Note the following given information. You will see questions in continuation to this, in the subsequent sections)

This situation has caused enough concern that the CEO is considering buying another plant, identical to the third, and using it permanently. Assuming that all 4 plants are on line, what production do you recommend to meet the current demand (5000, 8500, 10000)? In general, what can you say about any increased flexibility that the 4th plant might provide?

Let the number of barrels used by factory $A$, $B$, $C$ and $D$ are $x$, $y$, $z$ and $w$ respectively.

Then the system of linear equations will be

$$20x + 4y + 4z + 4w = 5000$$

$$10x + 14y + 5z + 5w = 8500$$

$$5x + 5y + 12z + 12w = 10000$$

The above system of linear equation has fewer equations than variables, hence it is *underdetermined* and cannot have a unique solution. In this case, there are either infinitely many solutions or no exact solution. We can solve it by keeping $w$ as constant and using [rref](http://linear.ups.edu/html/section-RREF.html) form to solve the system of linear equation.

To know about rref implementation in python refer [here](https://docs.sympy.org/latest/tutorial/matrices.html#rref).

In [85]:
import sympy as sy

# create symbol 'w'
w = sy.Symbol("w")
A_aug = sy.Matrix([[20, 4, 4, 5000-4*w],
                   [10, 14, 5, 8500-5*w],
                   [5, 5, 12, 10000-12*w]])
# show rref form
A_aug.rref()

(Matrix([
 [1, 0, 0,   195/4],
 [0, 1, 0,  1325/4],
 [0, 0, 1, 675 - w]]),
 (0, 1, 2))

From the above result, it can be seen that 4th plant will share the number of barrels required by the 3rd plant only, while the requirement of 1st and 2nd plant will remain unaffected.

### Calculate the amount of Paraffin supplied (1 point)

The company has just found a candle company that will buy its paraffin. Under the current conditions (i.e, after buying another plant) for demand (5000, 8500, 10000), how much can be supplied to them per day?

According to the problem statement, factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude oil, factory $B$ 5 gallons, and factory $C$ 2 gallons.

In [86]:
# YOUR CODE HERE
import sympy as sy

# create symbol 'w'
w = sy.Symbol("w")
A_aug = sy.Matrix([[20, 4, 4, 5000-4*w],
                   [10, 14, 5, 8500-5*w],
                   [5, 5, 12, 10000-12*w]])
# show rref form
row_echelon_form, _ = A_aug.rref()


In [87]:
row_echelon_form

Matrix([
[1, 0, 0,   195/4],
[0, 1, 0,  1325/4],
[0, 0, 1, 675 - w]])

In [88]:
paraffin_from_A = (195/4) * 3
paraffin_from_B = (1325/4) * 5
paraffin_from_C_D = (675) * 2
paraffin_supplied_per_day = paraffin_from_A + paraffin_from_B + paraffin_from_C_D
paraffin_supplied_per_day

3152.5

In [89]:
# if the disposal from D is different
parD = sy.Symbol('paraD')
paraffin_dispose = sy.Matrix([3,5,2,parD])
paraffin_supplied_w = row_echelon_form[:,-1].dot(paraffin_dispose[:-1]) + w*paraffin_dispose[-1]
print(paraffin_supplied_w)

paraD*w - 2*w + 6305/2


### Selling the first plant (1 point)

The management is also considering selling the first plant due to aging equipment and high workman's compensation costs for the state it is located in. They would like to know what this would do to their production capability. Specifically, they would like an example of a demand they could not meet with only plants 2 and 3, and also what effect having plant 4 has (recall it is identical to plant 3). They would also like an example of a demand that they could meet with just plants 2 and 3. Any general statements you could make here would be helpful.

Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

When considering only plants 2 and 3, and demand (5000, 8500, 10000) then we have

$$4y + 4z = 5000$$

$$14y + 5z = 8500$$

$$5y + 12z = 10000$$

**INFERENCE:** *The above system has 3 equations and 2 unknowns (y and z), which is overdetermined because number of linearly independent equations in the system are more than the unknowns*

In [90]:
b0 = sy.Symbol("b0")
b1 = sy.Symbol("b1")
b2 = sy.Symbol("b2")
A_aug = sy.Matrix([[4, 4, b0],
                   [14, 5, b1],
                   [5, 12,b2]])
# show rref form
echelon_form = A_aug.echelon_form()
print(echelon_form)

Matrix([[4, 4, b0], [0, -36, -14*b0 + 4*b1], [0, 0, 572*b0 - 112*b1 - 144*b2]])


In [91]:
echelon_form


Matrix([
[4,   4,                       b0],
[0, -36,            -14*b0 + 4*b1],
[0,   0, 572*b0 - 112*b1 - 144*b2]])

**Solution exists** if
 144b_2 = 572b_0 - 112b_1 (1)

 and

(14b_0 - 4b1)/36 >= 1  (2)

In [92]:
b_0 = 5000
b_1 = 8500
b_2 = (572/144)*b_0 - (112/144)*b_1
print(b_0,b_1,b_2)

5000 8500 13250.000000000002


[5000, 8500, 13250] demand can be met by plant B and C.

In [93]:
# YOUR CODE HERE
# For the first and second equations
A = np.array([[4, 4], [14, 5]])
b = np.array([5000, 8500])
X12 = solve(A,b)
X12

array([ 250., 1000.])

In [94]:
# For the second and third equations
A = np.array([[14, 5], [5, 12]])
b = np.array([8500, 13250])
X15 = solve(A,b)
X15

array([ 250., 1000.])

In [95]:
# For the first and third equations
A = np.array([[4, 4], [5, 12]])
b = np.array([5000, 13250])
X13 = solve(A,b)
X13

array([ 250., 1000.])

Taking 4th plant into consideration.
Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

Then for demand (5000, 8500, 10000) the system of linear equations will be

$$4y + 4z + 4w = 5000$$

$$14y + 5z + 5w = 8500$$

$$5y + 12z + 12w = 10000$$

Solve it using rref form.

if the demand is [5000, 8500, 13250], it can be met by plant B, C and D. We get the same condition as before, since we will arrive the set of equations if we write **z_prime = z + w**


In [96]:
# YOUR CODE HERE
import sympy as sy
w = sy.Symbol('w')
A_aug = sy.Matrix([[4, 4, 4*w, 5000],
                   [14, 5, 5*w, 8500],
                   [5, 12, 12*w, 10000]])
A_aug.rref()

(Matrix([
 [1, 0, 0, 0],
 [0, 1, w, 0],
 [0, 0, 0, 1]]),
 (0, 1, 3))

**INFERENCE:** *From above it can be seen that the above system of equation has no exact solution*
Reference:
https://people.math.osu.edu/husen.1/teaching/au2011/568/row_echelon_solutions.pdf

Now, changing demand to (6600, 5100, 3100) and solving the system of equation using rref form.

In [97]:
# YOUR CODE HERE
import sympy as sy
A_aug = sy.Matrix([[4, 4, 4, 6600],
                   [14, 5, 5, 5100],
                   [5, 12, 12, 3100]])
A_aug.rref()

(Matrix([
 [1, 0, 0, 0],
 [0, 1, 1, 0],
 [0, 0, 0, 1]]),
 (0, 1, 3))

In [98]:
# YOUR CODE HERE
import sympy as sy
A_aug = sy.Matrix([[4, 4, 5000-4*w],
                   [14, 5, 8500- 5*w],
                   [5, 12, 13250 - 12*w]])
A_aug.rref()

(Matrix([
 [1, 0,      250],
 [0, 1, 1000 - w],
 [0, 0,        0]]),
 (0, 1))

### Set rates for Products (1 point)

Company wants to set the rates of motor oil, diesel oil, and gasoline. For this purpose they have few suggestions given as follows:

* 100, 66, 102 Rupees per gallon,

* 104, 64, 100 Rupees per gallon,

* 102, 68, 98 Rupees per gallon, and

* 96, 68, 100 Rupees per gallon

for motor oil, diesel oil, and gasoline respectively.

Using matrix multiplication, find the rates which result in maximum total price.

Let $M$ denote the matrix such that rows represents different plants (A, B and C), columns represents different products (motor oil, diesel oil and gasoline) and each value represents production of that product from one barrel of crude oil for that plant.

$$M = \begin{bmatrix}
20 & 10 & 5 \\
4 & 14 & 5  \\
 4 & 5 & 12  
\end{bmatrix}$$

Also, $R$ is a matrix having different rates as its columns.

$$R = \begin{bmatrix}
100 & 104 & 102 & 96 \\
66 & 64 & 68 & 68  \\
102 & 100 & 98 & 100  
\end{bmatrix}$$

In [99]:
# YOUR CODE HERE
# Matrix multiplication
M = np.array([[20, 10, 5], [4, 14, 5], [4, 5, 12]])
R = np.array([[100, 104, 102, 96],
              [66, 64, 68, 70],
              [102, 100, 98, 100]])

M @ R

array([[3170, 3220, 3210, 3120],
       [1834, 1812, 1850, 1864],
       [1954, 1936, 1924, 1934]])

**EXPLAINATION:** *The resultant matrix contains total prices for each plant, taking different rates into consideration.*

The element (i: row,j: column) represents total price of plant i for a given price regime j

**NOW:** *In order to get the total price we need to sum up the prices for each plant (summing one column at a time) and take those rates for which the total price is maximum.*

In [100]:
# Total price for each rate
total_price = sum(M @ R)
print("Price: ", total_price)

# Get index of highest value in array
index = np.argmax(total_price)
print(index)

# Extract desired rates
rate = R[:,index]
print("Desired rate: ", rate)

Price:  [6958 6968 6984 6918]
2
Desired rate:  [102  68  98]


### Marginal Cost (1 point)

The total cost $C(x)$ in Rupees, associated with the production of $x$ gallons of gasoline is given by

$$C(x) = 0.005 x^3 – 0.02 x^2 + 30x + 5000$$

Find the marginal cost when $22$ gallons are produced, where, marginal cost means the instantaneous rate of change of total cost at any level of output.

**INTERPRETAION:**
*To get the marginal cost, we need to get the first derivartive of the equation and then place x=22 --> mathematically that will give us Rs. 36.38.*

*To code the following we will have to use lamda & derivative function from scipy package*

In [101]:
# YOUR CODE HERE
from scipy.misc import derivative

f = lambda x: 0.005 * x**3 - 0.02 * x**2 + 30 * x + 5000
x = 22
derivative(f, x)

<ipython-input-101-127325d391da>:6: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
  derivative(f, x)


36.38500000000022

### Marginal Revenue (1 point)

The total revenue in Rupees received from the sale of $x$ gallons of a motor oil is given by $$R(x) = 3x^2 + 36x + 5.$$

Find the marginal revenue, when $x = 28$, where, marginal revenue means the rate of change of total revenue with respect to the number of items sold at an instant.

In [102]:
# YOUR CODE HERE
from scipy.misc import derivative

f = lambda x: 3 * x**2 + 36 * x + 5
x = 28
derivative(f,x)

<ipython-input-102-6652035a74a4>:6: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
  derivative(f,x)


204.0

### Pouring crude oil in tank (1 point)

In a cylindrical tank of radius 10 meter, crude oil is being poured at the rate of 314 cubic meter per hour. Then find

* the rate at which the height of crude oil is increasing in the tank, and
* the height of crude oil in tank after 2 hours.

**CHANGE/REVISIT THIS (ANUSH)**

let q be the rate at which the ht. of the tank is increasing

the the height (h) of crude oil in cylinder, increasing with time t is

--> h = q * t / area

vol. of crude oil inside cyl. is

--> V = (pi)(r^2)h

and then, the rate of change of volume is

-->first derivative(V) wrt t = (pi)(r^2)q



In [103]:
# Solution-1

import numpy as np
import sympy as sym

# radius
r = 10

# time symbol
t = sym.Symbol('t')

# height rate symbol
q = sym.Symbol('m')

# height of tank
h = q * t

# volume of tank
v = np.pi * r**2 * h

# derivative(v,t)
# the rate of change of volume
sym.Derivative(v, t).doit()

314.159265358979*m

In [104]:
# Given volume flow rate = 314 cubic meter/hr in the question,
# solve for q
q = solve([[314.16]], [[314]])
print("Rate at which height of tank is increasing: ", q)

Rate at which height of tank is increasing:  [[0.99949071]]


**INFERENCE:** *the rate at which the height of the tank is increasing is 1 m/hr*

In [105]:
# now, let's calculate the ht. after two hrs
t = 2
h_2 = q * t
print("Height of crude oil in tank after 2 hours: ", h_2)

Height of crude oil in tank after 2 hours:  [[1.99898141]]


In [106]:
# Solution -2
import sympy as sy
import math as m
# Define symbolic variables

# Given values
radius = 10
dv_dt = 314
#calculate (dv_dt/dv_dh) = dh_dt since by chain rule dv/dt = (dv/dh)*(dh/dt)
h = sy.symbols('h')
r = sy.symbols('r')
V = m.pi*r**2*h
dv_dh = sy.diff(V,h)
# the rate at which the height of crude oil is increasing in the tank
print(dv_dt/dv_dh.subs(r,10))
# the height of crude oil in tank after 2 hours.
dh_dt = dv_dt/dv_dh.subs(r,10)
t = sy.symbols('t')
height = sy.integrate(dh_dt,(t,0,2))
print(height)

0.999493042617103
1.99898608523421


In [107]:
# Solution-3
from math import pi
pourRate = 314              # cubic meter per hour
r_tank = 10                 # meter
A_tank = pi* r_tank**2      # meter square
del_h= pourRate/A_tank      # meter per hour
print('The height of the crude oil is increasing at {} m per hour'.format(round(del_h,2)))
print('The height of crude oil in the tank is {} m after 2 hours.'.format(round(del_h*2,2)))

The height of the crude oil is increasing at 1.0 m per hour
The height of crude oil in the tank is 2.0 m after 2 hours.
